pkg/kubectl/cmd/util/factory.go

```diff
@@ -257,7 +261,41 @@ func NewFactory(optionalClientConfig clientcmd.ClientConfig) *Factory {
 				}
 				return kubectl.MakeLabels(t.Spec.Selector), nil
 			case *extensions.Deployment:
+				selector, err := unversioned.LabelSelectorAsSelector(t.Spec.Selector)
+				if err != nil {
+					return "", fmt.Errorf("failed to convert label selector to selector: %v", err)
+				}
+				return selector.String(), nil
+			default:
+				gvk, err := api.Scheme.ObjectKind(object)
+				if err != nil {
+					return "", err
+				}
+				return "", fmt.Errorf("cannot extract pod selector from %v", gvk)
+			}
+		},
+		MapBasedSelectorForObject: func(object runtime.Object) (string, error) {
+			// TODO: replace with a swagger schema based approach (identify pod selector via schema introspection)
+			switch t := object.(type) {
+			case *api.ReplicationController:
 				return kubectl.MakeLabels(t.Spec.Selector), nil
+			case *api.Pod:
+				if len(t.Labels) == 0 {
+					return "", fmt.Errorf("the pod has no labels and cannot be exposed")
+				}
+				return kubectl.MakeLabels(t.Labels), nil
+			case *api.Service:
+				if t.Spec.Selector == nil {
+					return "", fmt.Errorf("the service has no pod selector set")
+				}
+				return kubectl.MakeLabels(t.Spec.Selector), nil
+			case *extensions.Deployment:
```

Comment:
Can you explain the differences between `MapBasedSelectorForObject` and `PodSelectorForObject` (why do we need to separate them)? Is it because `ParseLabels` in pkg/kubectl/generate.go can only recognize and convert labels/selectors like `app=nginx` to `map[string]string` but not those like `app in (nginx)`?